### PR TITLE
#73 move security information into environment variables

### PR DESCRIPTION
**Pull request by davidradl**
Signed-off-by: David Radley <david_radley@uk.ibm.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,6 +79,20 @@ It is possible for multiple localServerName environment variables to be specifie
...
 
+#### The security environment variables.
+
+Environment variables used to secure communications between the presentation server and the view servers
+*EGERIA_PRESENTATIONSERVER_EGERIA_CA* This is the ca option as defined in the [node documentation](https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options).  
+*EGERIA_PRESENTATIONSERVER_EGERIA_PFX* This is the pfx option as defined in the [node documentation](https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options) 
+*EGERIA_PRESENTATIONSERVER_EGERIA_PFX_PASSPHRASE* This is the passphrase option as defined in the [node documentation](https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options) 
+
+Environment variables used to secure communications between the browser and the presention server
+*EGERIA_PRESENTATIONSERVER_EGERIA_BROWSER_CERT* This is the file location of the certificate to be used to secure communications between the browser and the presentation server
```
Comment (planetf1) [inline README.md:90]: It's good the variables are unique though the names are quite lengthy

File: README.md [hunk 2]
```diff
@@ -76,6 +79,20 @@ It is possible for multiple localServerName environment variables to be specifie
 Example:
 EGERIA_PRESENTATIONSERVER_SERVER_aaa={"remoteServerName":"cocoView1","remoteURL":"https://localhost:9443"}
 
+#### The security environment variables.
+
+Environment variables used to secure communications between the presentation server and the view servers
+*EGERIA_PRESENTATIONSERVER_EGERIA_CA* This is the ca option as defined in the [node documentation](https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options).
```
Comment (planetf1) [inline README.md:85]: ++ LIke the idea to just reference the node docs.

File: cra-server/.env_sample [hunk 3]
```diff
@@ -7,4 +7,20 @@
...
+EGERIA_PRESENTATIONSERVER_EGERIA_PFX=ssl/keystore.p12
+# This is the passphrase option as defined in the node documentation https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options  
+EGERIA_PRESENTATIONSERVER_EGERIA_PFX_PASSPHRASE=egeria
+
+# Environment variables used to secure communications between the browser and the presentation server
+
+# This is the file location of the certificate used to secure communications between the browser and the presentation server
+EGERIA_PRESENTATIONSERVER_EGERIA_BROWSER_CERT=ssl/keys/server.cert
+# This is the location of the file containing the key used to secure communications between the browser and the presentation server
+EGERIA_PRESENTATIONSERVER_EGERIA_BROWSER_PASSPHRASE=ssl/keys/server.key
```
Comment (planetf1) [inline cra-server/.env_sample:26]: The approach to setting the vars from the CLI needs clarifying. It's confusing currently.

I have a PR at https://github.com/odpi/egeria-react-ui/pull/84 which did this for the original var which I think is a slight improvement - but would need merging with these changes. Open to what order you merge in but I think that clarity is a must

**Review by planetf1 — CHANGES_REQUESTED** [review]
Can I suggest we move this to DRAFT status until/when the full stack has been tested complete with the coco pharma / container environments.

I'm unable to do this until the next release most likely and it's essential we don't break the UI in the interim

**Comment by planetf1** [comment]
The PR on egeria to update the certs is at https://github.com/odpi/egeria/pull/4722

At that point it would make sense to pull in any changes required for the UI

All the generated files - not just .p12 but PEM format too are checked in. I suggest the first pass is simply to copy. Longer term we need to figure out a cross-repo process for updating certs. Additionally docs will need to be updated in https://github.com/odpi/egeria/issues/4670

**Review by planetf1 — COMMENTED** [review]
Files for testing should be found in the samples we currently generate for egeria at open-metadata-resources/open-metadata-deployment/certificates

In here I think the relevant files are
 - EgeriaReactUIClient (.key.com & .p12) -> this is the key for the react ui acting as a client to the view server
  - EgeriaRootCA.p12 - this is what we refer to as the 'truststore' in base egeria and contains the root certificate authority certificate
   - 'egeria' - is the passphrase we use
   
   These are of course only sample files, created for development/demo use, and a real deployment should make use of their own keys. A sample script geared at developers using only open tools is in the 'gensamplecerts.sh' file
   
   So, I think we need to just check that those files can be consumed here -- there are various different formats for key stores / certs. This directory I mention includes both .p12 & .pem files so chose whichever works
  
 If that works, then  we need a PR against base egeria which will SET those environment variables, with the correct values/files , as part of the lab & egeria-base setup in k8s & compose

Also to consider is any additional documentation for someone picking up the UI and wanting to use with base egeria - I would be inclined to point them to the files above for sample certs rather than check them in - especially as it wouldn't surprise me if they get regenned in future. An entry in the readme showing a working example of the variables pointing at a local egeria source tree could be useful?

**Comment by planetf1** [comment]
One other thought - for the key used by presentation server these parms I think require the .pem format file, whilst for the ps<->egeria you are asking for the .p12 bundle .. I would still go ahead and see if this all works.

File: cra-server/functions/getAxiosInstance.js [hunk 4]
```diff
@@ -6,9 +6,11 @@ const https = require("https");
 const fs = require("fs");
 const path = require("path")
 const getServerInfoFromEnv = require('./getServerInfoFromEnv');
+const getSecurityInfoFromEnv = require('./getSecurityInfoFromEnv');
 
-const cert = fs.readFileSync(path.join(__dirname, '../../') + "ssl/keys/server.cert");
-const key = fs.readFileSync(path.join(__dirname, '../../') + "ssl/keys/server.key");
+const security = getSecurityInfoFromEnv();
+const cert = fs.readFileSync(path.join(__dirname, '../../') + security.browser_cert);
```
Comment (planetf1) [inline cra-server/functions/getAxiosInstance.js:12]: the docs need to clarify how the file can be specified in the given parameters
There should always be an option to specify an absolute path - for example we will likely need to MOUNT a filesystem from a kubernetes secret & it's much clearer/simple if that is specified in absolute terms as we use in egeria
(same comment for other parms)

The user would probably expect a relative path to be in relation to the directory where the server is run from - so if the above is needed to compensate for any directory switching done when when running node, then that's ok - perhaps just need to check for the user parm beginning with /

File: cra-server/router/routes.js [hunk 5]
```diff
@@ -11,23 +11,25 @@ const rateLimit = require("express-rate-limit");
...
 );
 // server for server authentication (so the server can trust us)
-const truststore = fs.readFileSync(
-  path.join(__dirname, "../../") + "ssl/truststore.p12"
+const ca = fs.readFileSync(
+  path.join(__dirname, "../../") + security.ca
 );
 
-passphrase = "egeria";
+passphrase= security.pfx_passphrase;
```
Comment (planetf1) [inline cra-server/router/routes.js:32]: so here you are using passphrase environment variable as a VALUE. 
Either is fine but should be consistent.

File: cra-server/router/routes.js [hunk 6]
```diff
@@ -233,9 +235,15 @@ router.get("/open-metadata/admin-services/*", (req, res) => {
     method: "get",
     url: urlRoot + incomingPath,
     httpsAgent: new https.Agent({
+<<<<<<< HEAD
```
Comment (planetf1) [inline cra-server/router/routes.js:238]: failed merge here

File: cra-server/router/routes.js (hunk 5 above)
Comment (planetf1) [reply cra-server/router/routes.js:32]: .. and more occurrences in file later on

File: cra-server/router/routes.js (hunk 6 above)
Comment (planetf1) [reply cra-server/router/routes.js:238]: and more later on

File: cra-server/.env_sample (hunk 3 above)
Comment (davidradl) [reply cra-server/.env_sample:26]: I have merged in 84, I had not pushed up all the code - sorry about that.

File: cra-server/functions/getAxiosInstance.js (hunk 4 above)
Comment (davidradl) [reply cra-server/functions/getAxiosInstance.js:12]: Agreed - I will change

File: cra-server/router/routes.js (hunk 6 above)
Comment (davidradl) [reply cra-server/router/routes.js]: Yes I havd not pushed up the last changes which addresses these. Sorry for the confusion

File: cra-server/router/routes.js (hunk 5 above)
Comment (davidradl) [reply cra-server/router/routes.js:32]: @planetf1 SorryI don't understand. Is there something inconsistent?

Comment (planetf1) [reply cra-server/router/routes.js:32]: Yes, on line 25/26 I see:
```
# This is the location of the file containing the key used to secure communications between the browser and the presentation server
EGERIA_PRESENTATIONSERVER_EGERIA_BROWSER_PASSPHRASE=ssl/keys/server.key
```
So I got the impression there the value was a file - whereas here the passphrase is referred to as a value?

**Comment by planetf1** [comment]
@davidradl I think this needs to stay open?

**Pull request closed without merging**